src/client.rs

```diff
@@ -303,47 +303,48 @@ impl Client {
     ///   then synchronize the local data with the server and return records, otherwise
-    ///   return an empty list.
+    ///   return an error.
     ///
     /// Note: with the [`DummyStorage`], any call to `.get()` will trigger a synchronization.
     ///
     /// Note: with `sync_if_empty` as `false`, if `.sync()` is never called then `.get()` will
-    /// always return an empty list.
+    /// always return an error.
     ///
     /// # Errors
     /// If an error occurs while fetching or verifying records, a [`ClientError`] is returned.
     pub fn get(&mut self) -> Result<Vec<Record>, ClientError> {
         let storage_key = self._storage_key();
 
         debug!("Retrieve from storage with key={:?}", storage_key);
-        let stored_bytes: Vec<u8> = self
-            .storage
-            .retrieve(&storage_key)
-            // TODO: surface errors. See #79
-            .unwrap_or_default();
-        let stored: Option<Collection> = serde_json::from_slice(&stored_bytes).unwrap_or(None);
-
-        match stored {
-            Some(collection) => {
+        let read_result = self.storage.retrieve(&storage_key);
+
+        match read_result {
+            Ok(stored_bytes) => {
+                // Deserialize content of storage and surface error if fails.
+                let stored: Collection = serde_json::from_slice(&stored_bytes).map_err(|err| {
+                    StorageError::ReadError(format!("cannot deserialize collection: {}", err))
+                })?;
+                // Verify signature of stored data (*optional*)
                 if !self.trust_local {
                     debug!("Verify signature of local data.");
-                    self.verifier.verify(&collection)?;
+                    self.verifier.verify(&stored)?;
                 }
 
-                Ok(collection.records)
+                Ok(stored.records)
             }
-            None => {
+            Err(err) => {
+                // If storage is empty, go on with sync() (*optional*)
                 if self.sync_if_empty {
-                    debug!("Synchronize data, without knowning which timestamp to expect.");
-                    let collection = self.sync(None)?;
-                    return Ok(collection.records);
+                    if let StorageError::KeyNotFound { .. } = err {
+                        debug!("Synchronize data, without knowning which timestamp to expect.");
+                        let collection = self.sync(None)?;
+                        return Ok(collection.records);
+                    }
                 }
-                // TODO: this empty list should be «qualified». Is it empty because never synced
-                // or empty on the server too. (see Normandy suitabilities).
-                debug!("Local data is empty or malformed.");
-                Ok(Vec::new())
+                // Otherwise, surface the error.
+                Err(err.into())
```

Comment:
These nested ifs could be simplified a bit by adding another arm to the outer `match` expression:
```
            // If storage is empty, go on with sync() (*optional*)
            Err(StorageError::KeyNotFound { .. }) if self.sync_if_empty => {
                debug!("Synchronize data, without knowning which timestamp to expect.");
                let collection = self.sync(None)?;
                Ok(collection.records)
            }
            // Otherwise, surface the error.
            Err(err) => Err(err.into()),
```
(I tried making this a GitHub suggestion, but it resulted in a diff that was all messed up.)

Reply:
Thanks I didn't know this syntax :)